### PR TITLE
Umbraco V8 support

### LINKED ISSUE #14: UmbracoExamine.PDF Support for Umbraco v8
Need to update the project to support v8

While we are doing this we should look at replacing iTextSharp (which requires a license) with PDFSharp (http://www.pdfsharp.net/) which is MIT licensed... if it works.

---
_This item has been added to our backlog [AB#1610](https://umbraco.visualstudio.com/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/1610)_

### PR DESCRIPTION
**Pull request by cleversolutions**
This is still a work in progress, but wanted to get the pull request started so we can start a discussion. There are a couple of problems with it at the moment. I've never build a nuget package before, hoping to get a bit of help there. The second issue, the write.lock file needs to be manually deleted before starting the website. Closes #14.

If anyone wants to try this out, you can download the nuget package from dropbox and drop it into your local repository.
https://www.dropbox.com/sh/gfjhcn262hgnj9e/AABpjfBhvFrPum7ofq9665kua?dl=0

### CODE REVIEW

**Comment by cleversolutions** [comment]
I've build a nuget package now, installed the package into a new U8 website, added the PDFs from the test project to the site and was able to successfully index them.

**Comment by cleversolutions** [comment]
I force the removal of the left over locks similar to what is done in ExamineComponent. It may be gross, hoping someone with more knowledge of Examine can comment on that.

**Comment by cleversolutions** [comment]
@Shazwazza this request is ready for review.

**Comment by cleversolutions** [comment]
I think I still have to add a component to index media items as they are added. Currently I think it requires an index rebuild to collect new content. I will have to do some more testing to be sure.

**Comment by cleversolutions** [comment]
Would it be better rather than include PDFExtractor.cs in this pull request to add https://www.nuget.org/packages/PdfTextract/ as a dependency?

**Comment by cleversolutions** [comment]
I added some comments to the code.

**Comment by Shazwazza** [comment]
Thanks for the work so far! We will get to reviewing this shortly but will most likely be in 1 or 2 weeks.

**Comment by cleversolutions** [comment]
Thanks Shannon. I’m in no rush. I’m going to be away all summer so I likely won’t be able to look at this again for another 8 weeks.

**Comment by bergmania** [comment]
Hi @cleversolutions.

Nice work - I have reviewed the code and made some small changes.
https://github.com/umbraco/UmbracoExamine.PDF/commit/2c6a8f37f12ec6ed245fefe5f7726ee87115c1fd

- Overall, I did not like that you used the `PDFExtractor` due the GPL license. It´s pretty simple to implement our own. - I did that
- I replaced some of the magic strings with the constants from Umbraco Core
- I Added some tests. Sadly one of the test pdf-files is not supported by PDFSharp 😢 
- I introduced an interface to the PdfTextExtractor, because there could be more versions. 
  - It is pretty easy to implement a ITextSharp version, or PDFBox version, if you have license to these. 
- I updated the version of PDFSharp to use the netstandard version

What do you think about my suggested changes?

**Comment by bergmania** [comment]
BTW, I agree with you about that you need to handle events, when pdf-files are added or removed from the system, the index need to be up-to-date.

**Comment by cleversolutions** [comment]
Thanks for the changes, they all look great. I've been curious about the difference between ILogger and IProfilingLogger but haven't had a chance to explore it. Your suggestion to introduce an interface to the extractor is excellent. I will be away for the next 8 weeks, so I will have to finish the changes to update the index when files are added/removed in the fall.

**Comment by cleversolutions** [comment]
Is there anything I need to do to include your changes in this PR?

**Comment by bergmania** [comment]
@cleversolutions, if you are happy with the changes, you can merge the branch `feature/support-v8-and-use-pdfsharp` into your branch.

**Comment by cleversolutions** [comment]
Thanks @bergmania. I've merged your changes and pushed in a very hastily coded feature to add/remove files to the index. Tested has been very minimal on this last change. On that note, I'm now gone for 8 weeks.

File: src/UmbracoExamine.PDF/PdfIndexPopulator.cs [hunk 1]
```diff
@@ -32,12 +33,99 @@ public PdfIndexPopulator(IMediaService mediaService, IPdfIndexValueSetBuilder me
...
+
+        /// <summary>
+        /// If pdfs were restored form recycle bin, re-index them
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void MediaService_Moved(IMediaService sender, Umbraco.Core.Events.MoveEventArgs<IMedia> e)
+        {
+            var restored = e.MoveInfoCollection
+                .Where(m => m.OriginalPath.StartsWith("-1")) //recyclebin
```
Comment (bergmania) [inline src/UmbracoExamine.PDF/PdfIndexPopulator.cs]: Media's recycle bin is `-21`.

```suggestion
                .Where(m => m.OriginalPath.StartsWith(Constants.System.RecycleBinMediaString))
```

File: src/UmbracoExamine.PDF/PdfIndexPopulator.cs [hunk 2]
```diff
@@ -32,12 +33,99 @@ public PdfIndexPopulator(IMediaService mediaService, IPdfIndexValueSetBuilder me
...
+        public PdfIndexPopulator(int? parentId, IMediaService mediaService, IPdfIndexValueSetBuilder mediaValueSetBuilder, IExamineManager examineManager)
         {
             _parentId = parentId;
             _mediaService = mediaService;
             _mediaValueSetBuilder = mediaValueSetBuilder;
+            _examineManager = examineManager;
             RegisterIndex("PDFIndex");
+
+            //Tie into various events one media saved/trashed/deleted/restored
+            Umbraco.Core.Services.Implement.MediaService.Saved += MediaService_Saved;
```
Comment (bergmania) [inline src/UmbracoExamine.PDF/PdfIndexPopulator.cs]: These should be handled by the component in initialize and not in a constructor.

File: src/UmbracoExamine.PDF/PdfIndexPopulator.cs [hunk 3]
```diff
@@ -32,12 +33,99 @@ public PdfIndexPopulator(IMediaService mediaService, IPdfIndexValueSetBuilder me
...
+        private void MediaService_Saved(IMediaService sender, Umbraco.Core.Events.SaveEventArgs<Umbraco.Core.Models.IMedia> e)
+        {
+            AddToIndex(e.SavedEntities);
+        }
+
+        /// <summary>
+        /// Add any media that is a pdf to the PDFIndex
+        /// </summary>
+        /// <param name="media"></param>
+        private void AddToIndex(IEnumerable<IMedia> media)
```
Comment (bergmania) [inline src/UmbracoExamine.PDF/PdfIndexPopulator.cs]: I think this should be `public` and used from the component..

Comment (bergmania) [reply src/UmbracoExamine.PDF/PdfIndexPopulator.cs]:
```suggestion
        public void AddToIndex(IEnumerable<IMedia> media)
```

File: src/UmbracoExamine.PDF/PdfIndexPopulator.cs [hunk 4]
```diff
@@ -32,12 +33,99 @@ public PdfIndexPopulator(IMediaService mediaService, IPdfIndexValueSetBuilder me
...
+        private void MediaService_Deleted(IMediaService sender, Umbraco.Core.Events.DeleteEventArgs<IMedia> e)
+        {
+            RemoveFromIndex(e.DeletedEntities);
+        }
+
+        /// <summary>
+        /// if any of the media is a pdf remove it from the PDFIndex
+        /// </summary>
+        /// <param name="media"></param>
+        private void RemoveFromIndex(IEnumerable<IMedia> media)
```
Comment (bergmania) [inline src/UmbracoExamine.PDF/PdfIndexPopulator.cs]: I think this should be `public` and used from the component
```suggestion
        public void RemoveFromIndex(IEnumerable<IMedia> media)
```

File: src/UmbracoExamine.PDF/PdfIndexPopulator.cs [hunk 5]
```diff
@@ -32,12 +33,99 @@ public PdfIndexPopulator(IMediaService mediaService, IPdfIndexValueSetBuilder me
...
+        {
+            if (_examineManager.TryGetIndex("PDFIndex", out var index))
```
Comment (bergmania) [inline src/UmbracoExamine.PDF/PdfIndexPopulator.cs]: `"PDFIndex"` is used a lot of places now.. Should be saved in a `const`

File: src/UmbracoExamine.PDF/PdfIndexPopulator.cs [hunk 6]
```diff
@@ -32,12 +33,99 @@ public PdfIndexPopulator(IMediaService mediaService, IPdfIndexValueSetBuilder me
...
+            {
+                var ids = media.Where(m => m.GetValue<string>("umbracoExtension") == "pdf").Select(m => m.Id.ToString());
```
Comment (bergmania) [inline src/UmbracoExamine.PDF/PdfIndexPopulator.cs]: `"pdf"` is used multiple places, and should be saved in a `const`

File: src/UmbracoExamine.PDF/PdfIndexPopulator.cs (hunk 3 above)
Comment (bergmania) [reply src/UmbracoExamine.PDF/PdfIndexPopulator.cs]: The method is almost identical to `RemoveFromIndex`
We could handle the common stuff in a single method, that takes a delegate that removes or add's to the index

**Comment by bergmania** [comment]
Hi @cleversolutions I fixed the review stuff and completed the PR.

I close this PR and continues on my branch (including all you commits)

Thanks for you contributions. You can see the entire PR here:
https://github.com/umbraco/UmbracoExamine.PDF/pull/17

**Pull request closed without merging**